jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp

```diff
@@ -0,0 +1,204 @@
+/*******************************************************************\
+
+Module: Unit tests for Java-object-factory's treatment of tag types
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <java-testing-utils/load_java_class.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/java_object_factory.h>
+#include <langapi/mode.h>
+#include <testing-utils/use_catch.h>
+
+static bool is_decl_with_id(const exprt &expr, const irep_idt &id)
+{
+  if(expr.id() != ID_code || to_code(expr).get_statement() != ID_decl)
+    return false;
+
+  const typet &decl_type = to_code_decl(to_code(expr)).symbol().type();
+  return decl_type.id() == ID_struct_tag &&
+         to_struct_tag_type(decl_type).get_identifier() == id;
+}
+
+static bool is_decl_of_type(const exprt &expr, const typet &type)
+{
+  return expr.id() == ID_code && to_code(expr).get_statement() == ID_decl &&
+         to_code_decl(to_code(expr)).symbol().type() == type;
+}
+
+static bool
+contains(const exprt &expr, std::function<bool(const exprt &)> predicate)
```

Comment:
:pick: Could these be implemented with a `depth_begin()`/`depth_end()`? Then can use `std::any`?